src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -21,36 +21,55 @@ internal static class CompletionHelpers
         /// <param name="wordToComplete">The word to complete.</param>
         /// <param name="possibleCompletionValues">The possible completion values to iterate.</param>
         /// <param name="toolTipMapping">The optional tool tip mapping delegate.</param>
+        /// <param name="listItemTextMapping">The optional list item text mapping delegate.</param>
         /// <param name="resultType">The optional completion result type. Default is Text.</param>
-        /// <returns></returns>
+        /// <param name="matchStrategy">The optional match strategy delegate.</param>
+        /// <returns>List of matching completion results.</returns>
         internal static IEnumerable<CompletionResult> GetMatchingResults(
             string wordToComplete,
             IEnumerable<string> possibleCompletionValues,
             Func<string, string> toolTipMapping = null,
-            CompletionResultType resultType = CompletionResultType.Text)
+            Func<string, string> listItemTextMapping = null,
+            CompletionResultType resultType = CompletionResultType.Text,
+            MatchStrategy matchStrategy = null)
         {
             string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
-            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
+
+            matchStrategy ??= DefaultMatch;
 
             foreach (string value in possibleCompletionValues)
             {
-                if (pattern.IsMatch(value))
+                if (matchStrategy(value, wordToComplete))
                 {
                     string completionText = QuoteCompletionText(value, quote);
+                    string toolTip = toolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = listItemTextMapping?.Invoke(value) ?? value;
 
-                    string listItemText = value;
-
-                    yield return new CompletionResult(
-                        completionText,
-                        listItemText,
-                        resultType,
-                        toolTip: toolTipMapping is null
-                            ? listItemText
-                            : toolTipMapping(value));
+                    yield return new CompletionResult(completionText, listItemText, resultType, toolTip);
                 }
             }
         }
 
+        internal delegate bool MatchStrategy(string value, string wordToComplete);
+
+        internal static readonly MatchStrategy LiteralMatch = (value, wordToComplete)
+            => value.StartsWith(
+                WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`")),
+                StringComparison.OrdinalIgnoreCase);
```

Comment:
I think the method must be explained why we need unescape and replace endings before ordinal compare. Please add a comment with example scenarios.

Reply:
Add XML documentation in 231f7bc0c0fe34ddbfab03f4ba1324dfe6a2a676

Reply:
It seems I skipped something... Why do we need WildcardPattern.Unescape()?

Reply:
I think this is an oversight or something which was working but I don't actually want to keep this. This is better: 8066ef5ce0c27474473425978dd639b69027feda